File: desktop/src/core/java/nativefont/NativeFontDesktop.java
package core.java.nativefont;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Pixmap;


import java.awt.BasicStroke;
import java.awt.Font;
import java.awt.FontFormatException;
import java.awt.FontMetrics;
import java.awt.Graphics2D;
import java.awt.RenderingHints;
import java.awt.Shape;
import java.awt.font.GlyphVector;
import java.awt.font.TextAttribute;
import java.awt.image.BufferedImage;
import java.io.BufferedInputStream;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.text.AttributedString;
import java.util.HashMap;

import javax.imageio.ImageIO;

import static javax.swing.UIManager.getColor;

/**
 * Created by tian on 2016/10/2.
 */

public class NativeFontDesktop implements NativeFontListener {
    private HashMap<String, Font> fonts = new HashMap<String, Font>();
    private HashMap<String, FontMetrics> metrics = new HashMap<String, FontMetrics>();
    private AttributedString as;

    public Pixmap getFontPixmap(String txt, NativeFontPaint vpaint) {
        Font font = getFont(vpaint);
        FontMetrics fm = metrics.get(vpaint.getName());
        int strWidth = fm.stringWidth(txt);
        int strHeight = fm.getAscent() + fm.getDescent();
        if (strWidth == 0) {
            strWidth = strHeight = vpaint.getTextSize();
        }
        BufferedImage bi = new BufferedImage(strWidth, strHeight,
                BufferedImage.TYPE_4BYTE_ABGR);
        Graphics2D g = bi.createGraphics();
        g.setRenderingHint(RenderingHints.KEY_ANTIALIASING,
                RenderingHints.VALUE_ANTIALIAS_ON);
        g.setFont(font);
        if (vpaint.getStrokeColor() != null) {
            // 描边
            GlyphVector v = font.createGlyphVector(fm.getFontRenderContext(),
                    txt);
            Shape shape = v.getOutline();
            g.setColor(getColor(vpaint.getColor()));
            g.translate(0, fm.getAscent());
            g.fill(shape);
            g.setStroke(new BasicStroke(vpaint.getStrokeWidth()));
            g.setColor(getColor(vpaint.getStrokeColor()));
            g.draw(shape);
        } else if (vpaint.getUnderlineText() == true) {
            // 下划线
            AttributedString as = new AttributedString(txt);
            as.addAttribute(TextAttribute.FONT, font);
            as.addAttribute(TextAttribute.UNDERLINE, TextAttribute.UNDERLINE_ON);
            g.setColor(getColor(vpaint.getColor()));
            g.drawString(as.getIterator(), 0, fm.getAscent());
        } else if (vpaint.getStrikeThruText() == true) {
            // 删除线
            AttributedString as = new AttributedString(txt);
            as.addAttribute(TextAttribute.FONT, font);
            as.addAttribute(TextAttribute.STRIKETHROUGH,
                    TextAttribute.STRIKETHROUGH_ON);
            g.setColor(getColor(vpaint.getColor()));
            g.drawString(as.getIterator(), 0, fm.getAscent());
        } else {
            // 普通
            g.setColor(getColor(vpaint.getColor()));
            g.drawString(txt, 0, fm.getAscent());
        }

        Pixmap pixmap = null;

        try {
            ByteArrayOutputStream buffer = new ByteArrayOutputStream();
            ImageIO.write(bi, "png", buffer);

            pixmap = new Pixmap(buffer.toByteArray(), 0, buffer.toByteArray().length);
            buffer.close();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return pixmap;
    }

    private Font getFont(NativeFontPaint vpaint) {
        boolean isBolo = vpaint.getFakeBoldText() || vpaint.getStrokeColor() != null;
        Font font = fonts.get(vpaint.getName());
        if (font == null) {
            if (vpaint.getTTFName().equals("")) {
                font = new Font("", isBolo ? Font.BOLD : Font.PLAIN, vpaint.getTextSize());
            } else {
                try {
                    ByteArrayInputStream in = new ByteArrayInputStream(Gdx.files.internal(vpaint.getTTFName() + (vpaint.getTTFName()
                            .endsWith(".ttf") ? "" : ".ttf")).readBytes());
                    BufferedInputStream fb = new BufferedInputStream(in);
                    font = Font.createFont(Font.TRUETYPE_FONT, fb).deriveFont(Font.BOLD, vpaint.getTextSize());
                    fb.close();
                    in.close();
                } catch (FontFormatException e) {
                    e.printStackTrace();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            fonts.put(vpaint.getName(), font);
            BufferedImage bi = new BufferedImage(1, 1, BufferedImage.TYPE_4BYTE_ABGR);
            Graphics2D g = bi.createGraphics();
            g.setFont(font);
            FontMetrics fm = g.getFontMetrics();
            metrics.put(vpaint.getName(), fm);
        }
        return font;
    }

}

File: desktop/src/core/java/nativefont/NativeFontDesktop.kt
package core.java.nativefont

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Pixmap
import java.awt.*
import java.awt.font.TextAttribute
import java.awt.image.BufferedImage
import java.io.BufferedInputStream
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.text.AttributedString
import java.util.*
import javax.imageio.ImageIO
import javax.swing.UIManager

class NativeFontDesktop : NativeFontListener {
    private val fonts = HashMap<String, Font?>()
    private val metrics = HashMap<String, FontMetrics>()
    override fun getFontPixmap(txt: String, vpaint: NativeFontPaint): Pixmap {
        val font = getFont(vpaint)
        val fm = metrics[vpaint.name]
        var strWidth = fm!!.stringWidth(txt)
        var strHeight = fm.ascent + fm.descent
        if (strWidth == 0) {
            strHeight = vpaint.textSize
            strWidth = strHeight
        }
        val bi = BufferedImage(strWidth, strHeight,
                BufferedImage.TYPE_4BYTE_ABGR)
        val g = bi.createGraphics()
        g.setRenderingHint(RenderingHints.KEY_ANTIALIASING,
                RenderingHints.VALUE_ANTIALIAS_ON)
        g.font = font
        when {
            vpaint.strokeColor != null -> { // 描边
                val v = font!!.createGlyphVector(fm.fontRenderContext,
                        txt)
                val shape = v.outline
                g.color = UIManager.getColor(vpaint.color)
                g.translate(0, fm.ascent)
                g.fill(shape)
                g.stroke = BasicStroke(vpaint.strokeWidth.toFloat())
                g.color = UIManager.getColor(vpaint.strokeColor)
                g.draw(shape)
            }
            vpaint.underlineText -> { // 下划线
                val `as` = AttributedString(txt)
                `as`.addAttribute(TextAttribute.FONT, font)
                `as`.addAttribute(TextAttribute.UNDERLINE, TextAttribute.UNDERLINE_ON)
                g.color = UIManager.getColor(vpaint.color)
                g.drawString(`as`.iterator, 0, fm.ascent)
            }
            vpaint.strikeThruText -> { // 删除线
                val `as` = AttributedString(txt)
                `as`.addAttribute(TextAttribute.FONT, font)
                `as`.addAttribute(TextAttribute.STRIKETHROUGH,
                        TextAttribute.STRIKETHROUGH_ON)
                g.color = UIManager.getColor(vpaint.color)
                g.drawString(`as`.iterator, 0, fm.ascent)
            }
            else -> { // 普通
                g.color = UIManager.getColor(vpaint.color)
                g.drawString(txt, 0, fm.ascent)
            }
        }
        lateinit var pixmap: Pixmap
        try {
            val buffer = ByteArrayOutputStream()
            ImageIO.write(bi, "png", buffer)
            pixmap = Pixmap(buffer.toByteArray(), 0, buffer.toByteArray().size)
            buffer.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return pixmap
    }

    private fun getFont(vpaint: NativeFontPaint): Font? {
        val isBolo = vpaint.fakeBoldText || vpaint.strokeColor != null
        var font = fonts[vpaint.name]
        if (font == null) {
            if (vpaint.tTFName == "") {
                font = Font("", if (isBolo) Font.BOLD else Font.PLAIN, vpaint.textSize)
            } else {
                try {
                    val `in` = ByteArrayInputStream(Gdx.files.internal(vpaint.tTFName + if (vpaint.tTFName
                                    .endsWith(".ttf")) "" else ".ttf").readBytes())
                    val fb = BufferedInputStream(`in`)
                    font = Font.createFont(Font.TRUETYPE_FONT, fb).deriveFont(Font.BOLD, vpaint.textSize.toFloat())
                    fb.close()
                    `in`.close()
                } catch (e: FontFormatException) {
                    e.printStackTrace()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
            fonts[vpaint.name] = font
            val bi = BufferedImage(1, 1, BufferedImage.TYPE_4BYTE_ABGR)
            val g = bi.createGraphics()
            g.font = font
            val fm = g.fontMetrics
            metrics[vpaint.name] = fm
        }
        return font
    }
}